### PR TITLE
INI: Travel routing

### PR DESCRIPTION
**Pull request by thomas-quadratic**
We initiate the Travel Routing project with this first PR.

## Features

We have 2 algorithms that work in FHE:

    1. one on an encrypted graph that finds shortest path between first and last vertex
    2. one where the client choose origin and destination, and the server returns the shortest path without learning the origin and destination

These algorithms are first working examples.
We include also a notebook to show how to use them:  nb_example.ipynb.

### CODE REVIEW

**Comment by haroune-mohammedi** [comment]
I just pushed the implementation of KNN to get the K nearest neighbors of a node

**Comment by aquint-zama** [comment]
relates https://github.com/zama-ai/bounty-program/issues/7

**Comment by thomas-quadratic** [comment]
Is also related to Issue #10

**Comment by andrei-stoian-zama** [comment]
Good job so far! While Dijkstra is a good idea, for now the compiler does not support loops, so you will not be able to increase the graph size by much (complexity in both code size and computation is O(N^3)). However, if it were possible to use only tensor computations in a less deep loop nest you could increase the graph size.

Second, could you explain what you use KNN for? You could also benefit from optimizing the KNN implementation by running the reduction step in parallel using a reduction tree. You would need to implement that by hand instead of using the reduce function of python.

File: TravelRouting/knn.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from concrete import fhe
+import numpy as np
+from functools import reduce
+
+def min_int(x: int, y: int) -> int:
+    """concrete-numpy doesn't yet support min, we have to implement one using
+    only supported operations"""
+    return (x + y - abs(x - y)) // 2
```
Comment (umut-sahin) [inline TravelRouting/knn.py:8]: This is using 2 PBS, you can do
```python
# min(x, y) == min(x - y, 0) + y
return fhe.univariate(lambda x_minus_y: x_minus_y if x_minus_y < 0 else 0)(x - y) + y
```
instead to do it in 1 PBS.

File: TravelRouting/knn.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    return reduce(min_int, a)
+
+def knn(distances):
+    """Return the index of the K minimum values (nearest neighbors) from the distances array
+    
+    The output is an array with zeros but ones at the indexes of the K min values
+    Note that the function could return more that K element if there are equal values
+    """
+    # if k is a parameter, we get: TypeError: 'Tracer' object cannot be interpreted as an integer
+    k = 2
```
Comment (umut-sahin) [inline TravelRouting/knn.py:21]: You can make it a parameter and use a lambda to compile to make it adjustable, but of course it cannot be a runtime value as we determine parameters from a static graph, so it cannot have unknown number of loops.

File: TravelRouting/knn.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        # Get the minimum value of the array
+        minimum = min_int_array(distances)
+        # Get a selector for the value
+        minimum_selector = distances == minimum
+        # Add the selected index to the result array
+        result = result | minimum_selector
+        # Get a selector for the remaining element
+        remaining_selector = (np.logical_not(minimum_selector.astype(bool))).astype(int)
+        # set the selected value to inf
+        distances = distances * remaining_selector + minimum_selector*inf
```
Comment (umut-sahin) [inline TravelRouting/knn.py:36]: Doing encrypted multiplications are highly inefficient as it'll generate 2 TLUs. Instead, I'd recommend doing something like https://github.com/zama-ai/concrete/blob/3456978c24df0a7397034c13c52a224ad5ed294c/frontends/concrete-python/examples/key-value-database/static-size.py#L68-L69, which is packing the boolean flag and the integer value in a single integer and doing a TLU there.

File: TravelRouting/routing.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+from concrete import fhe
+
+
+def take_min(x, y):
+    return (x + y - abs(x - y)) // 2
```
Comment (umut-sahin) [inline TravelRouting/routing.py:5]: Same as above.

File: TravelRouting/routing.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    distances[0] = 0
+    previous = [0] * N
+
+    for _ in range(1, N):
+        for v in range(N):
+            for u in range(N):
+                candidate = distances[u] + weights[u, v]
+                cond = (candidate < distances[v])
+                distances[v] = take_min(candidate, distances[v])
+                previous[v] = u * cond + previous[v] * (1 - cond)
```
Comment (umut-sahin) [inline TravelRouting/routing.py:33]: Same here.

**Comment by aquint-zama** [comment]
could be great to have either a working huggingface repo with the demo or instructions to make it run locally

**Comment by thomas-quadratic** [comment]
The branch is updated to include a streamlit demo app, that is also deployed on HuggingFace:
https://huggingface.co/spaces/tzamoj-quadratic/FHETravelRouting

**Comment by bcm-at-zama** [comment]
Hello @thomas-quadratic , please to meet you! And great to see you trying Concrete, it's awesome. 

So, my review would be:

### Positive
- good readme (I’ve not tried to install)
- quite didactic HF space: I like the steps and the fact that the flow is clear

### Negative
- I don’t think ‘restart’ needs to re-keygen in the space (minor thing)
- I am disapointed that the only FHE algorithm here is a TLU. It will not scale to lot of points.

Regarding the algorithm in FHE, it's more work. Maybe you can change the 'restart', which shouldn't be too hard (right?). For the FHE part, we'll see in the next milestones.

**Comment by bcm-at-zama** [comment]
Re-reading what I wrote: I miss some emphasize. The HF space is really cool, really really cool.

**Comment by RomanBredehoft** [comment]
Hello @thomas-quadratic , how are you ? Great to see that you were able to submit a working solution for this very interesting use case !

I have taken a look at your submission and here is what I have in mind : 

## Pros 
- I really liked the demo, it is very intuitive and works great ! Also, it's awesome that you have put it in a Hugging Face space
- The `Getting-Started.ipynb` notebook is very easy to follow
- The `README.md` file is also pretty clear, we directly understand the project's goal and installations steps are straightforward 


## Cons
- I also think the provided solution is pretty limited : I was expecting something like Dijstra in FHE or something similar, but here only a TLU is done in FHE, which is not very scalable. Also, I feel like the server should be able to return all nodes from a shortest path, and not only a single next node as it is currently done !
- there is little documentation in the code, which makes it pretty hard to follow 
- I was expecting more explanations on the problems encountered and how they were solved 


Overall, I think this makes a really great demo, I would probably just document a bit more the project. Regarding the FHE algorithm itself, as @bcm-at-zama said, we can see that in the next milestone !

Thanks a lot for your submission 😄

**Comment by thomas-quadratic** [comment]
Hi @bcm-at-zama @RomanBredehoft and nice to meet you too!

Glad that the streamlit app works great for you ! We will do the minor fixes that you mentioned (no key regeneration, code documentation, documenting milestones)

Regarding the algorithm, in a previous version, we had a shortest-path algo in FHE, but it was O(n^3) in the number of nodes. Also, it was inefficiently implemented using polynomials. Of what I understood, TFHE can do a TLU quite efficiently. We are nonetheless aware that we have a scaling problem.

Looking forward to discuss possible ways to improve the current implementation.
And thanks for your reviews!

**Comment by thomas-quadratic** [comment]
Hi, we have completed the changes we wanted to do:

1. Code refactoring and documentation so that it is easier to digest.
2. Single key generation / re-using the key after restarts.
3. Documenting in the readme our thoughts about improvements.

Would be great to discuss next steps as well !

**Comment by aquint-zama** [comment]
I'll merge this first version and let's discuss the next steps this week
